src/Exthost/Request.re

```diff
@@ -27,6 +28,93 @@ module Configuration = {
   };
 };
 
+module Decorations = {
+  type request = {
+    id: int,
+    handle: int,
+    uri: Uri.t,
+  };
+
+  let encodeRequest = request =>
+    Json.Encode.(
+      obj([
+        ("id", request.id |> int),
+        ("handle", request.handle |> int),
+        ("uri", request.uri |> Uri.encode),
+      ])
+    );
+
+  type decoration = {
+    priority: int,
+    bubble: bool,
+    title: string,
+    letter: string,
+    color: ThemeColor.t,
+  };
+
+  let (>>=::) = (fst, rest) => Json.Decode.uncons(rest, fst);
+
+  let decodeDecoration: Json.decoder(decoration) =
+    Json.Decode.(
+      int
+      >>=:: (
+        priority =>
+          bool
+          >>=:: (
+            bubble =>
+              string
+              >>=:: (
+                title =>
+                  string
+                  >>=:: (
+                    letter =>
+                      ThemeColor.decode
+                      >>=:: (
+                        color =>
+                          succeed({priority, bubble, title, letter, color})
+                      )
+                  )
+              )
+          )
+      )
+    );
```

Comment:
Ouch.. This is why we use `obj` 😆

Reply:
Ah, I see this is an array. In that case it might be better to use `index` with the pipeline API perhaps. Something like:

```suggestion
    Json.Decode.(
      Pipeline.(
        decode((priority, bubble, title, letter, color) => {priority, bubble, title, letter, color})
        |> custom(index(0, int))
        |> custom(index(1, bool))
        |> custom(index(2, string))
        |> custom(index(3, string))
        |> custom(index(4, ThemeColor.decode))
      )
    );
```

This should format better. Would be nice to have an `obj`-like decoder for arrays too though.

Reply:
Ouch, ya, this was brutal. I didn't know about the Pipeline API - that's a tremendous improvement - thanks @glennsl !